<a href="https://colab.research.google.com/github/FernandoColman/6375-projects/blob/main/NN_Knn_SVMs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# fetch data
from sklearn.datasets import fetch_openml

X, y = fetch_openml('mnist_784', version=1, return_X_y=True)

/usr/local/lib/python3.8/dist-packages/sklearn/datasets/_openml.py:932: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [3]:
# Split data into train and test
X = X / 255
X_train, X_test = X[:60000], X[60000:]
y_train, y_test = y[:60000], y[60000:]

In [ ]:
# SVM
from sklearn import svm

lc = svm.SVC(C=0.5, kernel='poly')
lc.fit(X_train, y_train)
y_pred = lc.predict(X_test)

# Testing Using Different Kernels/Penalization Parameter
### Linear 
> C value: 1.0  
> Execution Time: 5m 32s  
> Accuracy: 0.9404 

> C value: 1.5   
> Execution Time: 6m 43s   
> Accuracy: 0.9387   

> C value: 0.5   
> Execution Time: 5ms 24s   
> Accuracy: 0.943  

### RBF
> C value: 1.0  
> Execution Time: 6m 27s  
> Accuracy: 0.9792

> C value: 1.5  
> Execution Time: 7m 11s  
> Accuracy: 0.9821

> C value: 0.5  
> Execution Time: 7m 36s   
> Accuracy: 0.9759

### Polynomial
> C value: 1.0  
> Execution Time: 6ms 13s  
> Accuracy: 0.9771

> C value: 1.5  
> Execution Time: 3ms 34s   
> Accuracy: 0.9778

> C value: 0.5  
> Execution Time: 4m 47s 
> Accuracy: 0.9731

### Sigmoid
> C value: 1.0  
> Execution Time: 9m 47s   
> Accuracy: 0.7759




In [ ]:
# MLP
from sklearn.neural_network import MLPClassifier

#mlp = MLPClassifier(hidden_layer_sizes=(100),activation="relu",random_state=1).fit(X_train, y_train)
mlp = MLPClassifier(hidden_layer_sizes=(256,128,64,32),activation="relu", solver="lbfgs",random_state=1).fit(X_train, y_train)

y_pred = mlp.predict(X_test)
#print(mlp.score(X_test, y_test))

# Testing Using Different MLP Hyperparameters
### RELU 
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 1m 59s  
> Accuracy: 0.9811 

> Hidden Layers: (100)   
> Execution Time: 1m 21s  
> Accuracy: 0.9791  

> <b> Using LBFGS Solver </b>  
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 7m 44s  
> Accuracy: 0.9775   

> <b> Using SGD Solver </b>  
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 9m 22s  
> Accuracy: 0.9764   

### Identiy 
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 6m 42s  
> Accuracy: 0.9234 

> Hidden Layers: (100)   
> Execution Time: 2m 49s  
> Accuracy: 0.925 

### Logistic 
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 4m 25s  
> Accuracy: 0.9785 

> Hidden Layers: (100)   
> Execution Time: 2m 1s  
> Accuracy: 0.9771 

### Tanh 
> Hidden Layers: (256, 128, 64, 32)   
> Execution Time: 3m 19s  
> Accuracy: 0.9824   

> Hidden Layers: (100)   
> Execution Time: 1m 42s  
> Accuracy: 0.9783 

In [5]:
# KNN
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5, algorithm='brute', weights='distance')
knn.fit(X_train, y_train)
  

y_pred = knn.predict(X_test)
print(knn.score(X_test, y_test))

0.9691


# Testing Using Different KNN Hyperparameters
### Neighbors = 3
> Algorithm: Ball Tree     
> Execution Time: 17m 45s  
> Accuracy: 0.9705

> Algorithm: KD Tree     
> Execution Time: 20m 51s   
> Accuracy: 0.9705

> Algorithm: Brute    
> Execution Time: 49s   
> Accuracy:  0.9705

### Neighbors = 5
> Algorithm: Ball Tree     
> Execution Time: 17m 40s   
> Accuracy: 0.9688

> Algorithm: KD Tree     
> Execution Time: 20m 13s   
> Accuracy: 0.9678   

> Algorithm: Brute    
> Execution Time: 50s   
> Accuracy: 0.9829  

### Neighbors = 7
> Algorithm: Ball Tree     
> Execution Time: 18m 32s   
> Accuracy: 0.9688

> Algorithm: KD Tree     
> Execution Time: 22m 14s   
> Accuracy: 0.9690   

> Algorithm: Brute    
> Execution Time: 1m 17s   
> Accuracy:  0.9691

> Algorithm: Brute   
> Weights: Distance  
> Execution Time: 1m 18s   
> Accuracy: 0.9691

# Final Report 

In terms of error rate, the MLP Classifiers overall had a much better accuracy rate and relatively low execution times compared to the other classifiers. However, the RBF and Polynomial Kernels also performed well except with higher execution times. The very best combination of classifiers and hyperparameters in terms of accuracy was the KNN Classifier with 5 neighbors, the Brute algorithm, an execution time of 50s and an accuracy of 0.9829, however, the othe KNN instances had a really long execution time. 